pandas/core/sorting.py

```diff
@@ -285,6 +321,28 @@ def nargsort(
     return indexer
 
 
+def ensure_key_mapped(values, key: Optional[Callable]):
+    """
+    Applies a callable key function to elements in an Index subclass or
+    an ndarray. Uses index.map for index subclasses and ignores nan values
+    in ndarrays.
+
+    Parameters
+    ----------
+    values : Index subclass or ndarray
+    key : Optional[Callable], key to be called on every index or entry in ndarray.
+    """
+    if not key:
+        return values
+
+    result = key(values)
+    if len(result) != len(values):
```

Comment:
we don’t actually care about the types here right?
meaning if values is a Categoral
then result might be of might not be

Reply:
No. We could enforce types if we wanted, but it hasn't seemed necessary.

Reply:
@TomAugspurger can you explain what the `TODO: should ensure_key_mapped convert to an array?` was referring to? `ensure_key_mapped` needs to return an Index so we can call methods like `sortlevel`.

Reply:
> No. We could enforce types if we wanted, but it hasn't seemed necessary.

if we are not enforcing types (ok by me); then update the doc string (eg a Categorical is not required only a same length array-like)

Reply:
Ok. The only problem is that it can violate the return type – we don't type annotate it, but the documentation specifies that it returns a `Categorical` or None. But I updated the documentation.